src/librustc/traits/error_reporting.rs

```diff
@@ -487,13 +487,29 @@ impl<'a, 'gcx, 'tcx> InferCtxt<'a, 'gcx, 'tcx> {
                             } else {
                                 let trait_ref = trait_predicate.to_poly_trait_ref();
 
-                                let mut err = struct_span_err!(self.tcx.sess, span, E0277,
-                                    "the trait bound `{}` is not satisfied",
-                                    trait_ref.to_predicate());
-                                err.span_label(span, &format!("the trait `{}` is not implemented \
-                                                               for `{}`",
-                                                              trait_ref,
-                                                              trait_ref.self_ty()));
+                                let (post_message, pre_message) =
+                                    if let ObligationCauseCode::BuiltinDerivedObligation(ref data)
+                                        = obligation.cause.code {
+                                    let parent_trait_ref = self.resolve_type_vars_if_possible(
+                                        &data.parent_trait_ref);
+                                    (format!(" in `{}`", parent_trait_ref.0.self_ty()),
```

Comment:
One concern I have is that this only goes up one level, but there are similar cases where multiple levels are needed. For example, it'd be nice if this patch addressed this example: https://is.gd/BZjfP6

The current output is 

```
error[E0277]: the trait bound `*const u8: std::marker::Send` is not satisfied
  --> <anon>:16:5
   |
16 |     is_send::<Foo>();
   |     ^^^^^^^^^^^^^^ trait `*const u8: std::marker::Send` not satisfied
   |
   = note: `*const u8` cannot be sent between threads safely
   = note: required because it appears within the type `Baz`
   = note: required because it appears within the type `Bar`
   = note: required because it appears within the type `Foo`
   = note: required by `is_send`
```

and I think your suggested approach would be a big improvement:

```
error[E0277]: within `Foo`, the trait bound `*const u8: std::marker::Send` is not satisfied
  --> <anon>:16:5
   |
16 |     is_send::<Foo>();
   |     ^^^^^^^^^^^^^^ trait `*const u8: std::marker::Send` not satisfied
   |
   = note: `*const u8` cannot be sent between threads safely
   = note: required because it appears within the type `Baz`
   = note: required because it appears within the type `Bar`
   = note: required because it appears within the type `Foo`
   = note: required by `is_send`
```

But I believe your patch as is will print "within `Baz`" instead.

should be easy enough though -- just have to iterate on the cause chain rather than just going up one level.